l10n_it_fatturapa_in_pec/models/mail.py

```diff
@@ -0,0 +1,117 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Sergio Corato (https://efatto.it)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import logging
+import re
+import zipfile
+import os
+import base64
+from odoo import api, models
+from odoo.tools import config
+
+_logger = logging.getLogger(__name__)
+
+FATTURAPA_IN_REGEX = "^[A-Z]{2}[a-zA-Z0-9]{11,16}_[a-zA-Z0-9]{5}\.(xml|zip)"
+RESPONSE_MAIL_REGEX = '[A-Z]{2}[a-zA-Z0-9]{11,16}_[a-zA-Z0-9]{,5}_MT_' \
+                      '[a-zA-Z0-9]{,3}'
+
+
+class MailThread(models.AbstractModel):
+    _inherit = 'mail.thread'
+
+    @api.model
+    def message_route(self, message, message_dict, model=None, thread_id=None,
+                      custom_values=None):
+
+        if any("@pec.fatturapa.it" in x for x in [
+            message.get('Reply-To', ''),
+            message.get('From', ''),
+            message.get('Return-Path', '')
+        ]):
+            _logger.info("Processing FatturaPA PEC Invoice with Message-Id: "
+                         "{}".format(message.get('Message-Id')))
+            fatturapa_attachment_in = self.env['fatturapa.attachment.in']
+            fatturapa_regex = re.compile(FATTURAPA_IN_REGEX)
+            fatturapa_attachments = [x for x in message_dict['attachments']
+                                     if fatturapa_regex.match(x.fname)]
+            response_regex = re.compile(RESPONSE_MAIL_REGEX)
+            response_attachments = [x for x in message_dict['attachments']
+                                    if response_regex.match(x.fname)]
+            if response_attachments and fatturapa_attachments:
+                if len(response_attachments) > 1:
+                    _logger.info(
+                        'More than 1 message found in mail of incoming '
+                        'invoice')
+                message_dict['model'] = fatturapa_attachment_in._name
+                message_dict['record_name'] = message_dict['subject']
+                message_dict['res_id'] = 0
+                attachment_ids = self._message_post_process_attachments(
+                    message_dict['attachments'], [], message_dict)
+                for attachment in self.env['ir.attachment'].browse(
+                        [att_id for model, att_id in attachment_ids]):
+                    if fatturapa_regex.match(attachment.name):
+                        is_stored = True if attachment._storage() == 'file'\
+                            else False
+                        if is_stored:
+                            path = os.path.join(
+                                config['data_dir'], "filestore",
+                                self.env.cr.dbname)
+                            file_path = os.path.join(
+                                path, attachment.store_fname)
+                        else:
+                            pass  # todo prendi il file dal db?
```

Comment:
@eLBati ci sono degli esempi di apertura del file dal db?

Reply:
ho modificato, vedo che i dati sono sempre accessibili dall'attachment

Reply:
@sergiocorato la gestione è trasparente: gli attachment e le loro API restano invariati. La differenza, impostando il salvataggio nel DB, è che i file binari non vengono appunto salvati su file system. Quindi dovresti usare sempre i metodi di `ir.attachment` per leggere e scrivere. Poi ci pensa lui

Reply:
@eLBati ottimo, vedi il commit seguente, i dati li trovo in .datas direi

Reply:
Esatto

Reply:
ok, hai visto il commit https://github.com/OCA/l10n-italy/pull/636/commits/a2efc93105dbc82ab57ae1c2b7bdd769485adcc7 ?

Reply:
Mi pare vada bene

Reply:
ok